### DISCUSSION TITLE
How to use AppWindow

### DISCUSSION THREAD

**Discussion by ArchieCoder** [unoplatform/uno#18338] [General]
In the blog post of 5.4.5 https://platform.uno/blog/5-4/#:~:text=September%2026,%202024.%20We%20are%20introducing%20over%20290%20new%20additions, there is a mention of New AppWindow APIs to get/set the position and size as well as move, resize and show/hide windows of the app.

AppWindow does not compile in Skia.

<img width="592" alt="image" src="https://github.com/user-attachments/assets/be4d45e0-340f-4c60-9957-7cd60a9ae0c6">

Regarding WinUI, it does not work with the code below. WinUIEx by @dotMorten works.

    public MainPage()
    {
        InitializeComponent();

        Loaded += MainPage_Loaded;
        Unloaded += MainPage_Unloaded;
    }

    private void MainPage_Loaded(object sender, RoutedEventArgs e)
    {
        _appWindow = AppWindow.Create();

        RestoreWindowPosition();
    }

    private void MainPage_Unloaded(object sender, RoutedEventArgs e)
    {
        SaveWindowPosition();
    }

    private void SaveWindowPosition()
    {
        var settings = ApplicationData.Current.LocalSettings;
        settings.Values["WindowWidth"] = _appWindow.Size.Width;
        settings.Values["WindowHeight"] = _appWindow.Size.Height;
        settings.Values["WindowX"] = _appWindow.Position.X;
        settings.Values["WindowY"] = _appWindow.Position.Y;
    }

    private void RestoreWindowPosition()
    {
        var settings = ApplicationData.Current.LocalSettings;

        if (settings.Values.TryGetValue("WindowWidth", out var width) &&
            settings.Values.TryGetValue("WindowHeight", out var height) &&
            settings.Values.TryGetValue("WindowX", out var x) &&
            settings.Values.TryGetValue("WindowY", out var y))
        {
            var size = new SizeInt32 { Width = (int)width, Height = (int)height };
            var position = new PointInt32 { X = (int)x, Y = (int)y };
            _appWindow.Resize(size);
            _appWindow.Move(position);
        }
    }

**Comment by MartinZikmund** [comment]
@ArchieCoder the `AppWindow.Create()` method is not implemented yet, instead you can create a XAML `Microsoft.UI.Xaml.Window` and then use its `AppWindow` property, e.g.:

```csharp
var window = new Window();
window.AppWindow.SetPresenter(...);
```

To access the main (primary) window of your app, you can make the `Window` instance in `App.xaml.cs` accessible as a public property for example

**Reply by ArchieCoder** [reply]
The full screen state does not seem to work when restoring the window presenter or I am probably missing something. I created a sample project where you can see the behavior. 

[UnoAppWindow.zip](https://github.com/user-attachments/files/17211467/UnoAppWindow.zip)

https://github.com/user-attachments/assets/36f8d4cc-47fa-4426-8755-57171fa55454

`        MainWindow = builder.Window;
        MainWindow.Closed += OnMainWindowClosed;
        RestoreWindowPosition();

#if DEBUG
        MainWindow.EnableHotReload();
#endif
        MainWindow.SetWindowIcon();

        Host = await builder.NavigateAsync<Shell>();
    }

    private void OnMainWindowClosed(object sender, WindowEventArgs args)
    {
        SaveWindowPosition();
    }

    private void SaveWindowPosition()
    {
        var settings = ApplicationData.Current.LocalSettings;
        settings.Values["WindowWidth"] = MainWindow!.AppWindow.Size.Width;
        settings.Values["WindowHeight"] = MainWindow.AppWindow.Size.Height;
        settings.Values["WindowX"] = MainWindow.AppWindow.Position.X;
        settings.Values["WindowY"] = MainWindow.AppWindow.Position.Y;
        settings.Values["Presenter"] = (int)MainWindow.AppWindow.Presenter.Kind;
    }

    private void RestoreWindowPosition()
    {
        var settings = ApplicationData.Current.LocalSettings;

        if (settings.Values.TryGetValue("WindowWidth", out var width) &&
            settings.Values.TryGetValue("WindowHeight", out var height) &&
            settings.Values.TryGetValue("WindowX", out var x) &&
            settings.Values.TryGetValue("WindowY", out var y) &&
            settings.Values.TryGetValue("Presenter", out var p))
        {
            var size = new SizeInt32 { Width = (int)width, Height = (int)height };
            var position = new PointInt32 { X = (int)x, Y = (int)y };

            MainWindow!.AppWindow.Resize(size);
            MainWindow.AppWindow.Move(position);
            MainWindow.AppWindow.SetPresenter((AppWindowPresenterKind)p);
        }
    }`

**Comment by dotMorten** [comment]
@ArchieCoder careful with the code above. If your monitor layout changes your window could end up off-screen. Winuiex handles that and only restores for last known layout

**Reply by ArchieCoder** [reply]
Thanks for the hint. You convinced me to keep your library for WinUI. I do need something for Skia.
